hazelcast/src/main/java/com/hazelcast/util/executor/DiagnosticScheduledExecutor.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.util.executor;
+
+import com.hazelcast.logging.ILogger;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.Delayed;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.RunnableScheduledFuture;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static com.hazelcast.logging.Logger.getLogger;
+import static java.lang.Thread.currentThread;
+
+/**
+ * {@link java.util.concurrent.ScheduledExecutorService} stops a task silently by default if there is an exception.
+ * This class adds logging behaviour for exceptional cases by overriding {@link ScheduledThreadPoolExecutor#afterExecute}
+ * and {@link ScheduledThreadPoolExecutor#decorateTask} methods.
+ *
+ * @see {@link java.util.concurrent.ScheduledExecutorService#scheduleWithFixedDelay}
+ */
+public class DiagnosticScheduledExecutor extends ScheduledThreadPoolExecutor {
+
+    private final ILogger logger = getLogger(getClass());
+
+    public DiagnosticScheduledExecutor(int corePoolSize, ThreadFactory threadFactory) {
+        super(corePoolSize, threadFactory);
+    }
+
+    @Override
+    protected <V> RunnableScheduledFuture<V> decorateTask(Runnable runnable, RunnableScheduledFuture<V> task) {
```

Comment:
Why do the tasks need to be decorated with something that only provides a toString implementation?